src/libopensc/card-dtrust.c

```diff
@@ -213,10 +271,44 @@ dtrust_init(sc_card_t *card)
 
 	card->cla = 0x00;
 
-	card->drv_data = calloc(1, sizeof(struct dtrust_drv_data_t));
-	if (card->drv_data == NULL)
+	drv_data = calloc(1, sizeof(struct dtrust_drv_data_t));
+	if (drv_data == NULL)
 		return SC_ERROR_OUT_OF_MEMORY;
 
+	drv_data->pace = 0;
+	drv_data->can = 0;
+	drv_data->can_value = NULL;
+	drv_data->can_cache = 1;
+	drv_data->p15card = NULL;
+
+	drv_data->can_value = can_env = getenv("DTRUST_CAN");
```

Comment:
This buffer is not safe for later use. better use `getenv_s` or, at the very least, make a copy from the value.

Reply:
To be portable between Linux and Windows I decided to stay with `getenv()` here, but to copy the CAN value into a secure buffer. This also addresses your `sc_mem_secure_alloc()`-remark from above.